framework/py/flwr/serverapp/strategy.py

```diff
@@ -162,88 +162,121 @@ def start(
-                )
-                del replies_list
-                # Log training metrics and append to history
-                if agg_arrays is None or agg_metrics is None:
-                    break
-                log(INFO, "\t└──> Aggregated  MetricRecord: %s", agg_metrics)
-                results.train_metrics[current_round] = agg_metrics
-                results.arrays = agg_arrays
-                arrays = agg_arrays
-
-            # Configure evaluate, send messages and wait for replies
-            replies = grid.send_and_receive(
-                messages=self.configure_evaluate(
-                    current_round, arrays, evaluate_config, grid
-                ),
+            del train_messages
+
+            # Aggregate train
+            agg_arrays, agg_train_metrics = self.aggregate_train(
+                current_round,
+                list(train_replies),
+            )
+            del train_replies
+            
+            # Log training metrics and append to history
+            if agg_arrays is None or agg_train_metrics is None:
+                break
+            log(INFO, "\t└──> Aggregated MetricRecord: %s", agg_train_metrics)
+            result.train_metrics[current_round] = agg_train_metrics
+            result.arrays = agg_arrays
+            arrays = agg_arrays
+
+            # -----------------------------------------------------------------
+            # --- EVALUATION (LOCAL) ------------------------------------------
+            # -----------------------------------------------------------------
+
+            # Call strategy to configure evaluation round
+            evaluate_messages = self.configure_evaluate(
+                current_round,
+                arrays,
+                evaluate_config,
+                grid,
+            )
+
+            # Send messages and wait for replies
+            evaluate_replies = grid.send_and_receive(
+                messages=evaluate_messages,
                 timeout=timeout,
             )
-            replies_list = list(replies)
-            if replies_list:
-                # Aggregate evaluate
-                agg_metrics = self.aggregate_evaluate(current_round, replies_list)
-                if agg_metrics is None:
-                    break
-                log(INFO, "\t└──> Aggregated MetricRecord: %s", agg_metrics)
-                results.evaluate_metrics[current_round] = agg_metrics
-
-            # Centralised eval
-            if central_eval_fn:
-                log(INFO, "Central evaluation")
-                res = central_eval_fn(current_round, arrays)
+            del evaluate_messages
+
+            # Aggregate evaluate
+            agg_evaluate_metrics = self.aggregate_evaluate(
+                current_round,
+                list(evaluate_replies),
+            )
+            del evaluate_replies
+
+            # Log training metrics and append to history
+            if agg_evaluate_metrics is None:
+                break
```

Comment:
Do we want to end FL if there are no aggregation metrics? (this would be the case also if the run is configure to not sample nodes for evaluation). One idea is to not `break` and replace the above with:

```python
            if agg_metrics is not None:
                log(INFO, "\t└──> Aggregated MetricRecord: %s", agg_metrics)
                results.evaluate_metrics[current_round] = agg_metrics
```

Reply:
a similar logic could be followed for `aggregate_train()`